jpa/subsystem/src/main/java/org/jboss/as/jpa/container/EntityManagerCacheGlobal.java

```diff
@@ -0,0 +1,87 @@
+/*
+ *  JBoss, Home of Professional Open Source.
+ *  Copyright 2018, Red Hat, Inc., and individual contributors
+ *  as indicated by the @author tags. See the copyright.txt file in the
+ *  distribution for a full listing of individual contributors.
+ *
+ *  This is free software; you can redistribute it and/or modify it
+ *  under the terms of the GNU Lesser General Public License as
+ *  published by the Free Software Foundation; either version 2.1 of
+ *  the License, or (at your option) any later version.
+ *
+ *  This software is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ *  Lesser General Public License for more details.
+ *
+ *  You should have received a copy of the GNU Lesser General Public
+ *  License along with this software; if not, write to the Free
+ *  Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ *  02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+
+package org.jboss.as.jpa.container;
+
+import java.util.concurrent.ConcurrentLinkedDeque;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import javax.persistence.EntityManager;
+
+import org.jipijapa.plugin.spi.EntityManagerCache;
+
+/**
+ * EntityManagerCacheGlobal is a Global (per application deployment/per persistence unit cache.
+ *
+ * @author Scott Marlow
+ */
+public class EntityManagerCacheGlobal implements EntityManagerCache {
```

Comment:
Which part is not thread safe currently?  Are you concerned that the cache_size might not be exact?

Reply:
With the `put` method you could end up with more entries in the cache than the `max_size`. Imagine that a checking is blocking on the `cache_size.get()` then immediately after the lock is released another `cache_size.get()` returns the same number that may equal the `max_size`. Both could be cached and the `cache_size` incremented twice resulting int a value larger than `max_size`.

Reply:
There's a race in the `clear()` method too which is similar to above.

Reply:
The clear method should only be called at application undeploy time.  I added a comment to the SPI indicating why it should only be called at undeploy time.

I'm looking into the put method issue...

Reply:
In the put method, the cache_size.get() doesn't block, get() returns the current (volatile) value.  

The various threads that are calling this method at the same time, will loop in the call to cache_size.getAndIncrement(), as this call is atomic.  

I will update this method to eliminate the race but will still use (mostly) lock contention free datastructures for better concurrency, as this code is called a lot from multiple concurrent application threads.

Reply:
I pushed a concurrent test that showed failures with yesterdays 'racy' cache implementation, with todays correction, we are now passing the TransactionTestCase.testConcurrentAccess().
Thanks James for the great feedback, good improvements today!  I'll wait until you have a chance to review again.

Also, let me know if its okay for me to squash these commits into one.  I thought you might want to see the evolution of todays changes, so pushed separate commits. :)

Reply:
This still feels a bit racy to me. I really think it should look something more like:
```
package org.jboss.as.jpa.container;

import java.util.Collections;
import java.util.IdentityHashMap;
import java.util.Iterator;
import java.util.Set;
import javax.persistence.EntityManager;

import org.jboss.as.jpa.messages.JpaLogger;
import org.jipijapa.plugin.spi.EntityManagerCache;

/**
 * EntityManagerCacheGlobal is a Global (per application deployment/per persistence unit cache.
 * EntityManagerCacheGlobal is thread-safe.
 *
 * @author Scott Marlow
 */
public class EntityManagerCacheGlobal implements EntityManagerCache {

    private final Set<EntityManager> cache = Collections.newSetFromMap(new IdentityHashMap<>());
    private final int maxSize;

    /**
     * @param maxSize if greater than zero, will use bounded cache, otherwise cache can have unlimited # of EntityManagers.
     */
    public EntityManagerCacheGlobal(int maxSize) {
        this.maxSize = maxSize;
    }

    @Override
    public synchronized EntityManager get() {
        EntityManager result = null;
        final Iterator<EntityManager> iter = cache.iterator();
        while (iter.hasNext()) {
            result = iter.next();
            iter.remove();
            if (result.isOpen()) {
                JpaLogger.ROOT_LOGGER.tracef("Taking EntityManager %s to cache", result);
                break;
            }
        }
        return result;
    }

    @Override
    public synchronized void put(EntityManager entityManager) {
        if (maxSize <= 0 || cache.size() < maxSize) {
            if (entityManager.isOpen()) {
                if (cache.add(entityManager)) {
                    JpaLogger.ROOT_LOGGER.tracef("Added EntityManager %s to cache", entityManager);
                } else {
                    JpaLogger.ROOT_LOGGER.debugf("Failed to add EntityManager %s to cache", entityManager);
                }
            }
        } else {
            if (entityManager.isOpen()) {
                entityManager.close();
            }
        }
    }

    // NOTE: only expected to be called at application undeployment time.
    @Override
    public synchronized void clear() {
        // Drain and close
        final Iterator<EntityManager> iter = cache.iterator();
        while (iter.hasNext()) {
            final EntityManager em = iter.next();
            iter.remove();
            if (em.isOpen()) {
                em.close();
            }
        }
    }
}
```